### DISCUSSION TITLE
How to interact with several gate controllers?

### DISCUSSION THREAD

**Discussion by moazam1** [uhppoted/uhppoted#37] [Q&A] [answered]
Hi - Just checking if there is any possibility to connect several gate controllers located in different locations/regions with a single server? The server URL will be same for all the devices. Can we flash firmware, etc. of the device so that the device can talk to the server and registered in our database when connected first time?

**Comment by uhppoted** [comment]
Yes, it's straightforward and there's no need to flash firmware or do anything special because the controller protocol is _request-response_ (i.e. the controller responds to the server that sent the request). 

However, to receiving events from the controllers the controllers have to be configured to send the events to a specific server, using a _set-event-listener_ command (e.g. [CLI set-listener](https://github.com/uhppoted/uhppote-cli#set-listener)).

One thing to note - the protocol as currently implemented in the _uhppoted_ libraries uses UDP and by default requests are sent as a UDP broadcast, which means the server and controllers have to be on the same LAN segment. For controllers and servers on different LAN segments, it is necessary to configure the IP addresses of the controllers in the _uhppoted.conf_ configuration file [CONTROLLERS](https://github.com/uhppoted/uhppoted/blob/master/documentation/uhppoted.conf/uhppoted.conf.md#controllers) section.

**Comment by moazam1** [comment]
@uhppoted thank you for getting back to me quick. The main challenge we are facing here is that controllers do not have static IP. They have to communicate with the server and establish either a socket connection or something similar so that they can send/receive commands. Think it's doable?

**Reply by YBonline** [reply, marked as answer]
I have a bunch of different locations setup, my choice was to setup a VPN tunnel (Wireguard) between the different locations to be able to assign local IP addresses and avoid exposing anything directly to the internet. I have other systems (such as cameras) that I also did not want facing the internet directly or having to deal with tunneling, and Wireguard can be installed on many existing routers (so no extra hardware, even if a raspberry pi is cheap)

For my real "remote" locations, I have an OpenWRT router that takes a SIM card to provide internet and also connects to Wireguard to my server to get the information over in a local matter. On my locations with more permanent and complicated setups, I've been using UniFi Gateways with Wireguard on them as well, so this approach frequently wouldn't require any extra hardware, but is more configuration to setup.

Its important to note that these controllers have literally no security implemented on them, so your LAN setup is the only security your access control system gets... which should sound scary because it is. Ideally these should get segregated to their own VLAN to make sure no traffic you don't expect to can communicate with these controller cards, because anyone on a network with them can do things like unlock your doors if they are simply on your network.

**Reply by uhppoted** [reply]
Thanks YB - yup, a VPN is really the way to go! (And good point about the security)

**Comment by uhppoted** [comment]
To the best of my knowledge the controllers do not support DHCP lookup - typically they're configured with a static IP (using the [set-address](https://github.com/uhppoted/uhppote-cli#set-address) command. They also don't have the functionality required to connect to a server.

If you can't assign static addresses then one way to do it is to use UDP broadcast - the controllers will self assign an IPv4 address and will respond to requests addressed to their serial number. But that does mean they have to be reachable via UDP broadcast which means they have to be on the same LAN segment unless you are able to configure your network to forward UDP broadcasts (most routers don't).

Another way to do what you seem to have in mind is to install something like a _RaspberryPi_ close to the controller and have that act as a UDP relay (using e.g. [uhppoted-tunnel](https://github.com/uhppoted/uhppoted-tunnel) or something similar).
